File: mocks/auth-mock/src/main/java/no/nav/pensjon/vtp/auth/STSIssueResponseGenerator.java
package no.nav.pensjon.vtp.auth;

import static java.util.Collections.singletonList;

import java.security.Principal;
import java.util.HashMap;
import java.util.Map;

import javax.security.auth.callback.Callback;
import javax.security.auth.callback.CallbackHandler;

import org.apache.cxf.security.SecurityContext;
import org.apache.cxf.sts.StaticSTSProperties;
import org.apache.cxf.sts.operation.TokenIssueOperation;
import org.apache.cxf.sts.request.ReceivedToken;
import org.apache.cxf.sts.request.RequestRequirements;
import org.apache.cxf.sts.request.TokenRequirements;
import org.apache.cxf.sts.token.delegation.UsernameTokenDelegationHandler;
import org.apache.cxf.sts.token.provider.SAMLTokenProvider;
import org.apache.cxf.sts.token.provider.TokenProviderParameters;
import org.apache.cxf.ws.security.sts.provider.model.ObjectFactory;
import org.apache.cxf.ws.security.sts.provider.model.RequestSecurityTokenCollectionType;
import org.apache.cxf.ws.security.sts.provider.model.RequestSecurityTokenResponseCollectionType;
import org.apache.cxf.ws.security.sts.provider.model.RequestSecurityTokenResponseType;
import org.apache.cxf.ws.security.sts.provider.model.RequestSecurityTokenType;
import org.apache.wss4j.common.crypto.Crypto;
import org.apache.wss4j.common.ext.WSPasswordCallback;
import org.apache.wss4j.common.principal.CustomTokenPrincipal;
import org.springframework.stereotype.Component;

import no.nav.pensjon.vtp.felles.CryptoConfigurationParameters;

@Component
public class STSIssueResponseGenerator {

    private static final String USERNAME = "goddagmannøkseskaft";
    private final CryptoConfigurationParameters cryptoConfiguration;
    private final TokenIssueOperation issueOperation;

    public STSIssueResponseGenerator(Crypto crypto, final CryptoConfigurationParameters cryptoConfiguration) {
        this.cryptoConfiguration = cryptoConfiguration;
        this.issueOperation = createIssueOperation(initSTSProperties(crypto));
    }


    static class PasswordCallbackHandler implements CallbackHandler {

        private final CryptoConfigurationParameters cryptoConfigurationParameters;

        public PasswordCallbackHandler(CryptoConfigurationParameters cryptoConfigurationParameters) {
            this.cryptoConfigurationParameters = cryptoConfigurationParameters;
        }

        @Override
        public void handle(Callback[] callbacks) {
            for (Callback callback : callbacks) {
                if (callback instanceof WSPasswordCallback) { // CXF
                    WSPasswordCallback pc = (WSPasswordCallback) callback;
                    if (cryptoConfigurationParameters.getKeyAndCertAlias().equals(pc.getIdentifier())) {
                        pc.setPassword(cryptoConfigurationParameters.getKeystorePassword());
                        break;
                    }
                }
            }
        }
    }

    private static TokenIssueOperation createIssueOperation(StaticSTSProperties staticSTSProperties) {
        TokenIssueOperation tokenIssueOperation = new TokenIssueOperation() {
            @Override
            protected TokenProviderParameters createTokenProviderParameters(RequestRequirements requestRequirements,
                    Principal principal,
                    Map<String, Object> messageContext) {
                TokenProviderParameters providerParameters = super.createTokenProviderParameters(requestRequirements, principal, messageContext);
                TokenRequirements tokenRequirements = providerParameters.getTokenRequirements();
                tokenRequirements.setOnBehalfOf(null);
                tokenRequirements.setActAs(null);
                return providerParameters;
            }
        };

        // Add Token Provider
        tokenIssueOperation.setTokenProviders(singletonList(new SAMLTokenProvider()));

        // Add TokenDelegationHandler for onBehalfOf
        tokenIssueOperation.setDelegationHandlers(singletonList(
                new UsernameTokenDelegationHandler() {
                    @Override
                    public boolean canHandleToken(ReceivedToken delegateTarget) {
                        return true;
                    }
                }
                ));

        tokenIssueOperation.setStsProperties(staticSTSProperties);

        return tokenIssueOperation;
    }

    private Map<String, Object> createMessageContext(Principal principal) {
        Map<String, Object> messageContext = new HashMap<>();
        messageContext.put(
            SecurityContext.class.getName(),
            createSecurityContext(principal));
        return messageContext;
    }

    public RequestSecurityTokenResponseCollectionType buildRequestSecurityTokenResponseCollectionType(RequestSecurityTokenCollectionType requestCollection) {
        Principal principal = new CustomTokenPrincipal(USERNAME);
        Map<String, Object> messageContext = createMessageContext(principal);
        return issueOperation.issue(requestCollection, principal, messageContext);
    }

    /** Issue a token as part of collection */
    public RequestSecurityTokenResponseCollectionType buildRequestSecurityTokenResponseCollectionType(RequestSecurityTokenType request) {
        Principal principal = new CustomTokenPrincipal(USERNAME);
        Map<String, Object> messageContext = createMessageContext(principal);
        return issueOperation.issue(request, principal, messageContext);
    }

    /** Issue a single token */
    public RequestSecurityTokenResponseType buildRequestSecurityTokenResponseType(RequestSecurityTokenType request) {
        Principal principal = new CustomTokenPrincipal(USERNAME);
        Map<String, Object> messageContext = createMessageContext(principal);
        return issueOperation.issueSingle(request, principal, messageContext);
    }

    /** Issue a single token */
    public RequestSecurityTokenResponseType buildRequestSecurityTokenResponseType(String tokenType){
        ObjectFactory of = new ObjectFactory();
        RequestSecurityTokenType request2 = of.createRequestSecurityTokenType();
        request2.getAny().add(of.createTokenType(tokenType));
        request2.getAny().add(of.createRequestType("http://docs.oasis-open.org/ws-sx/ws-trust/200512/Issue"));

        Principal principal = new CustomTokenPrincipal(USERNAME);
        Map<String, Object> messageContext = createMessageContext(principal);
        return issueOperation.issueSingle(request2, principal, messageContext);
    }

    private static SecurityContext createSecurityContext(final Principal p) {
        return new SecurityContext() {
            @Override
            public Principal getUserPrincipal() {
                return p;
            }

            @Override
            public boolean isUserInRole(String role) {
                return false;
            }
        };
    }

    private StaticSTSProperties initSTSProperties(Crypto crypto) {
        // Add STSProperties object
        String alias = cryptoConfiguration.getKeyAndCertAlias();
        StaticSTSProperties stsProperties = new StaticSTSProperties();
        stsProperties.setEncryptionCrypto(crypto);
        stsProperties.setSignatureCrypto(crypto);
        stsProperties.setCallbackHandler(new PasswordCallbackHandler(cryptoConfiguration));
        stsProperties.setSignatureUsername(alias);
        stsProperties.setIssuer("VTP");
        return stsProperties;
    }
}

File: mocks/auth-mock/src/main/java/no/nav/pensjon/vtp/auth/STSIssueResponseGenerator.kt
package no.nav.pensjon.vtp.auth

import no.nav.pensjon.vtp.felles.CryptoConfigurationParameters
import org.apache.cxf.security.SecurityContext
import org.apache.cxf.sts.StaticSTSProperties
import org.apache.cxf.sts.operation.TokenIssueOperation
import org.apache.cxf.sts.request.ReceivedToken
import org.apache.cxf.sts.request.RequestRequirements
import org.apache.cxf.sts.token.delegation.UsernameTokenDelegationHandler
import org.apache.cxf.sts.token.provider.SAMLTokenProvider
import org.apache.cxf.sts.token.provider.TokenProviderParameters
import org.apache.cxf.ws.security.sts.provider.model.*
import org.apache.wss4j.common.crypto.Crypto
import org.apache.wss4j.common.ext.WSPasswordCallback
import org.apache.wss4j.common.principal.CustomTokenPrincipal
import org.springframework.stereotype.Component
import java.security.Principal
import java.util.*
import javax.security.auth.callback.Callback
import javax.security.auth.callback.CallbackHandler

@Component
class STSIssueResponseGenerator(crypto: Crypto, private val cryptoConfiguration: CryptoConfigurationParameters) {
    private val issueOperation: TokenIssueOperation

    internal class PasswordCallbackHandler(private val cryptoConfigurationParameters: CryptoConfigurationParameters) :
        CallbackHandler {
        override fun handle(callbacks: Array<Callback>) {
            for (callback in callbacks) {
                if (callback is WSPasswordCallback) { // CXF
                    if (cryptoConfigurationParameters.keyAndCertAlias == callback.identifier) {
                        callback.password = cryptoConfigurationParameters.keystorePassword
                        break
                    }
                }
            }
        }
    }

    private fun createMessageContext(principal: Principal): Map<String, Any> {
        val messageContext: MutableMap<String, Any> = HashMap()
        messageContext[SecurityContext::class.java.name] = createSecurityContext(principal)
        return messageContext
    }

    fun buildRequestSecurityTokenResponseCollectionType(requestCollection: RequestSecurityTokenCollectionType?): RequestSecurityTokenResponseCollectionType {
        val principal: Principal = CustomTokenPrincipal(USERNAME)
        val messageContext = createMessageContext(principal)
        return issueOperation.issue(requestCollection, principal, messageContext)
    }

    /** Issue a token as part of collection  */
    fun buildRequestSecurityTokenResponseCollectionType(request: RequestSecurityTokenType?): RequestSecurityTokenResponseCollectionType {
        val principal: Principal = CustomTokenPrincipal(USERNAME)
        val messageContext = createMessageContext(principal)
        return issueOperation.issue(request, principal, messageContext)
    }

    /** Issue a single token  */
    fun buildRequestSecurityTokenResponseType(request: RequestSecurityTokenType?): RequestSecurityTokenResponseType {
        val principal: Principal = CustomTokenPrincipal(USERNAME)
        val messageContext = createMessageContext(principal)
        return issueOperation.issueSingle(request, principal, messageContext)
    }

    /** Issue a single token  */
    fun buildRequestSecurityTokenResponseType(tokenType: String?): RequestSecurityTokenResponseType {
        val of = ObjectFactory()
        val request2 = of.createRequestSecurityTokenType()
        request2.any.add(of.createTokenType(tokenType))
        request2.any.add(of.createRequestType("http://docs.oasis-open.org/ws-sx/ws-trust/200512/Issue"))
        val principal: Principal = CustomTokenPrincipal(USERNAME)
        val messageContext = createMessageContext(principal)
        return issueOperation.issueSingle(request2, principal, messageContext)
    }

    private fun initSTSProperties(crypto: Crypto) = StaticSTSProperties().apply {
        encryptionCrypto = crypto
        signatureCrypto = crypto
        callbackHandler = PasswordCallbackHandler(
            cryptoConfiguration
        )
        signatureUsername = cryptoConfiguration.keyAndCertAlias
        issuer = "VTP"
    }

    companion object {
        private const val USERNAME = "goddagmannøkseskaft"
        private fun createIssueOperation(staticSTSProperties: StaticSTSProperties): TokenIssueOperation {
            val tokenIssueOperation: TokenIssueOperation = object : TokenIssueOperation() {
                override fun createTokenProviderParameters(
                    requestRequirements: RequestRequirements,
                    principal: Principal,
                    messageContext: Map<String, Any>
                ): TokenProviderParameters {
                    val providerParameters =
                        super.createTokenProviderParameters(requestRequirements, principal, messageContext)
                    val tokenRequirements = providerParameters.tokenRequirements
                    tokenRequirements.onBehalfOf = null
                    tokenRequirements.actAs = null
                    return providerParameters
                }
            }

            // Add Token Provider
            tokenIssueOperation.tokenProviders = listOf(SAMLTokenProvider())

            // Add TokenDelegationHandler for onBehalfOf
            tokenIssueOperation.delegationHandlers = listOf(
                object : UsernameTokenDelegationHandler() {
                    override fun canHandleToken(delegateTarget: ReceivedToken): Boolean {
                        return true
                    }
                }
            )
            tokenIssueOperation.setStsProperties(staticSTSProperties)
            return tokenIssueOperation
        }

        private fun createSecurityContext(p: Principal): SecurityContext {
            return object : SecurityContext {
                override fun getUserPrincipal(): Principal {
                    return p
                }

                override fun isUserInRole(role: String): Boolean {
                    return false
                }
            }
        }
    }

    init {
        issueOperation = createIssueOperation(initSTSProperties(crypto))
    }
}
